core/src/main/java/edu/ucr/cs/riple/core/global/GlobalAnalyzerImpl.java

```diff
@@ -0,0 +1,217 @@
+    System.out.println("Analyzing downstream dependencies completed!");
+  }
+
+  /**
+   * Retrieves the corresponding {@link MethodImpact} to a fix.
+   *
+   * @param fix Target fix.
+   * @return Corresponding {@link MethodImpact}, null if not located.
+   */
+  @Nullable
+  private MethodImpact fetchStatus(Fix fix) {
+    if (!fix.isOnMethod()) {
+      return null;
+    }
+    OnMethod onMethod = fix.toMethod();
+    int predictedHash = MethodImpact.hash(onMethod.method, onMethod.clazz);
+    Optional<MethodImpact> optional =
+        this.methods.get(predictedHash).stream()
+            .filter(m -> m.node.location.equals(onMethod))
+            .findAny();
+    return optional.orElse(null);
+  }
+
+  /**
+   * Returns the effect of applying a fix on the target on downstream dependencies.
+   *
+   * @param fix Fix targeting an element in target.
+   * @param fixesLocation Location in target that will be annotated as {@code @Nullable}.
+   * @return Effect on downstream dependencies.
+   */
+  private int effectOnDownstreamDependencies(Fix fix, Set<Location> fixesLocation) {
+    MethodImpact status = fetchStatus(fix);
+    if (status == null) {
+      return 0;
+    }
+    int individualEffect = status.getEffect();
+    // Some triggered errors might be resolved due to fixes in the tree, and we should not double
+    // count them.
+    List<Error> triggeredErrors = status.getTriggeredErrors();
+    long resolvedErrors =
+        triggeredErrors.stream()
+            .filter(error -> fixesLocation.contains(error.nonnullTarget))
+            .count();
+    return individualEffect - (int) resolvedErrors;
+  }
+
+  @Override
+  public int computeLowerBoundOfNumberOfErrors(Set<Fix> tree) {
+    Set<Location> fixesLocation =
+        tree.stream().map(f -> f.change.location).collect(Collectors.toSet());
+    OptionalInt lowerBoundEffectOfChainOptional =
+        tree.stream().mapToInt(fix -> effectOnDownstreamDependencies(fix, fixesLocation)).max();
+    if (lowerBoundEffectOfChainOptional.isEmpty()) {
+      return 0;
+    }
+    return lowerBoundEffectOfChainOptional.getAsInt();
+  }
+
+  @Override
+  public int computeUpperBoundOfNumberOfErrors(Set<Fix> tree) {
+    Set<Location> fixesLocation =
+        tree.stream().map(f -> f.change.location).collect(Collectors.toSet());
+    return tree.stream().mapToInt(fix -> effectOnDownstreamDependencies(fix, fixesLocation)).sum();
+  }
+
+  @Override
+  public ImmutableSet<OnParameter> getImpactedParameters(Set<Fix> fixTree) {
+    return fixTree.stream()
+        .filter(Fix::isOnMethod)
+        .flatMap(
+            fix -> {
+              MethodImpact status = fetchStatus(fix);
+              return status == null ? Stream.of() : status.getImpactedParameters().stream();
+            })
+        .collect(ImmutableSet.toImmutableSet());
+  }
+
+  @Override
+  public List<Error> getTriggeredErrors(Fix fix) {
+    // We currently only store impact of methods on downstream dependencies.
```

Comment:
Might be worth having an issue to track this for arguments/fields and link it here. Though I think that's a fairly low priority issue to solve right now, it's worth documenting the limitation and making sure we remember it's there 🙂

Reply:
Created an issue #66. :)